File: app/src/main/assets/resources/js/HalHash.js
/*class HalHash {
    constructor() {this.version="0.0.1";}
    addition(num) {
        if (num.length<2){return BigInteger.BigInt(num,10);}
        var h=parseInt(num.length/2);
        return BigInteger.add(BigInteger.add(BigInteger.BigInt(num.substring(0,h),10),BigInteger.BigInt(num.substring(h),10)),BigInteger.BigInt("1",10));
    }
    
    Str2Hash(str,length) {
        var out=[];
        if (str.length==0){out=[-1];}
        for (var i=0;i<str.length;i++){out.push(str.charCodeAt(i));}
        return this.Bin2Hash(new Uint8Array(out),length);
    }
    
    Bin2Hex(buffer) {
        return [...buffer]
        .map(x => x.toString(16).padStart(2, '0'))
        .join('');
    }

    Bin2Hash(bytes,length) {
        if (length<4){console.error("Minimum length=4")}
        var o=BigInteger.BigInt("0x"+this.Bin2Hex(bytes),16);
        var isNeg=o<0;
        o=o.toString(10);
        if (isNeg){o=o.substring(1);}
        if(o.length<4){
            var h=parseInt(o)/3+2.0;
            var j=4-o.length;
            while (parseInt(h).toString().length!=j) {
                while (parseInt(h).toString().length<j) {h=Math.pow(h,3);}
                while (parseInt(h).toString().length>j) {h/=3;}
            }
            o+=parseInt(h).toString();
        }
        var p=parseInt(o.length/4);
        var a=BigInteger.BigInt(o.substring(0,p),10);
        if (isNeg){a=BigInteger.sub(BigInteger.unaryMinus(a),BigInteger.BigInt("2"));}
        else {a=BigInteger.add(a,BigInteger.BigInt("2"));}
        var ad=a.toString(10).length;
        var b=BigInteger.BigInt(o.substring(p,p*2),10);
        if (isNeg){b=BigInteger.sub(BigInteger.unaryMinus(b),BigInteger.BigInt("2"));}
        else {b=BigInteger.add(b,BigInteger.BigInt("2"));}
        var bd=b.toString(10).length;
        var c=BigInteger.BigInt(o.substring(p*2,p*3),10);
        if (isNeg){c=BigInteger.sub(BigInteger.unaryMinus(c),BigInteger.BigInt("2"));}
        else {c=BigInteger.add(c,BigInteger.BigInt("2",10));}
        var cd=c.toString(10).length;
        var d=BigInteger.BigInt(o.substring(p*3),10);
        if (isNeg){d=BigInteger.sub(BigInteger.unaryMinus(d),BigInteger.BigInt("2"));}
        else {d=BigInteger.add(d,BigInteger.BigInt("2",10));}
        var dd=d.toString(10).length;
        var pd=parseInt(length/4);
        var od=length-pd*3;
        while (!(a.toString(16).length==pd && b.toString(16).length==pd && c.toString(16).length==pd && d.toString(16).length==od)) {
            while (a.toString(16).length<pd){a=BigInteger.exponentiate(a,BigInteger.BigInt("3",10));}
            while (a.toString(16).length>pd){a=this.addition(a.toString(10));}
            while (b.toString(16).length<pd){b=BigInteger.exponentiate(b,BigInteger.BigInt("3",10));}
            while (b.toString(16).length>pd){b=this.addition(b.toString(10));}
            while (c.toString(16).length<pd){c=BigInteger.exponentiate(c,BigInteger.BigInt("3",10));}
            while (c.toString(16).length>pd){c=this.addition(c.toString(10));}
            while (d.toString(16).length<od){d=BigInteger.exponentiate(d,BigInteger.BigInt("3",10));}
            while (d.toString(16).length>od){d=this.addition(d.toString(10));}
        }
        return a.toString(16)+b.toString(16)+c.toString(16)+d.toString(16);
    }
}*/

/*class HalHash {
    constructor() {this.version="0.0.2";}
    
    
    Str2Hash(str,length,count=64) {
        //var out=[];
        //if (str.length==0){out=[-1];}
        //for (var i=0;i<str.length;i++){out.push(str.charCodeAt(i));}
        //return this.Bin2Hash(new Uint8Array(out),length,count);
        return this.Bin2Hash((new TextEncoder("utf-8")).encode(str),length,count);
    }
    
    Bin2Hex(buffer) {
        return [...buffer]
        .map(x => x.toString(16).padStart(2, '0'))
        .join('');
    }
    
    array_toHex(arr) {
        var out="";
        for (var i=0;i<arr.length;i++) {
            out+=arr[i].toString(16);
        }
        return out;
    }
    
    getHalfInt(i,h) {
        var temp=i.toString();
        temp=h==0?temp.slice(0,parseInt(temp.length/2)):temp.slice(parseInt(temp.length/2),temp.length);
        return temp==""?BigInteger.BigInt(0):BigInteger.BigInt(temp);
    }

    Bin2Hash(bytes,length,count=64, maxIteration=100000) {
        //if (bytes.length==1){bytes=new Uint8Array([bytes[0],0]);}
        count=length<count?length:count;
        count=bytes.length<count?bytes.length:count;
        var sizeD=parseInt(bytes.length/count);
        var d=[];
        var i,b;
        var g=BigInteger.multiply(BigInteger.BigInt(count),BigInteger.BigInt(length));
        for (i=0;i<count;i++) {
            d.push(BigInteger.multiply(BigInteger.BigInt("0x"+this.Bin2Hex(bytes.slice(i*sizeD,i*sizeD+sizeD)),16),g));
        }
        if (bytes.length%count) {
            d[count-1]=BigInteger.multiply(BigInteger.BigInt("0x"+this.Bin2Hex(bytes.slice(count*sizeD-sizeD,bytes.length)),16),g);
        }
        
        var p=parseInt((length-count)/2);
        for (i=0;i<p;i++) {
            d.push(i+2);
            for (b=0;b<count;b++) {
                d[d.length-1]=BigInteger.add(BigInteger.multiply(d[b],d[d.length-1]),BigInteger.multiply(g,i+1));
            }
        }
        count=d.length;
        var temp=d.slice();
        
        g=BigInteger.add(this.getHalfInt(g,0),this.getHalfInt(g,1));
        for (i=0;i<count;i++) {
            for (b=0;b<count;b++) {
                //d[b]=BigInteger.multiply(d[b],BigInteger.add(BigInteger.add(d[b],BigInteger.multiply(temp[i],temp[i])),BigInteger.multiply(temp[i],g)));
                d[b]=BigInteger.add(d[b],BigInteger.add(BigInteger.multiply(temp[i],temp[i]),BigInteger.multiply(temp[i],g)));
            }
        }
        var h="",lc=count-1;
        var mc=0;
        while (1) {
            h=this.array_toHex(d);
            if (h.length<length) {
                for (i=0;i<count;i++) {
                    if (i==lc) {
                        //d[i]=BigInteger.add(BigInteger.multiply(this.getHalfInt(d[i],0),this.getHalfInt(d[i],1)),g);
                        d[i]=BigInteger.add(BigInteger.add(d[i],d[i]),g)
                    } else {
                        temp=BigInteger.add(this.getHalfInt(d[i],0),this.getHalfInt(d[i],1));
                        d[i]=BigInteger.BigInt(temp.toString()+(BigInteger.add(this.getHalfInt(d[i+1],0),this.getHalfInt(d[i+1],1))).toString());
                        d[i+1]=BigInteger.BigInt((BigInteger.add(this.getHalfInt(d[i+1],0),this.getHalfInt(d[i+1],1))).toString()+temp.toString());
                    }
                }
            } else if (h.length>length) {
                if (mc>maxIteration) {
                    h=h.substring(0,length);
                    break;
                }
                
                for (i=0;i<count;i++) {
                    if (i==lc) {
                        d[i]=BigInteger.add(BigInteger.add(this.getHalfInt(d[i],0),this.getHalfInt(d[i],1)),g);
                    } else {
                        temp=this.getHalfInt(d[i],1);
                        d[i]=BigInteger.add(this.getHalfInt(d[i],0),this.getHalfInt(d[i+1],1));
                        d[i+1]=BigInteger.add(this.getHalfInt(d[i+1],0),temp);
                    }
                }
            } else {break;}
            mc++;
        }
        return h;
    }
}*/

/*class HalHash {
    constructor() {this.version="0.0.2";}
    
    
    Str2Hash(str,length,count=64) {
        return this.Bin2Hash((new TextEncoder("utf-8")).encode(str),length,count);
    }
    
    Bin2Hex(buffer) {
        return [...buffer]
        .map(x => x.toString(16).padStart(2, '0'))
        .join('');
    }
    
    array_toHex(arr) {
        var out="";
        for (var i=0;i<arr.length;i++) {
            out+=arr[i].toString(16);
        }
        return out;
    }
    
    getHalfInt(i,h) {
        var temp=i.toString();
        temp=h==0?temp.slice(0,parseInt(temp.length/2)):temp.slice(parseInt(temp.length/2),temp.length);
        return temp==""?BigInt(0):BigInt(temp);
    }

    Bin2Hash(bytes,length,count=64, maxIteration=100000) {
        //if (bytes.length==1){bytes=new Uint8Array([bytes[0],0]);}
        count=length<count?length:count;
        count=bytes.length<count?bytes.length:count;
        var sizeD=parseInt(bytes.length/count);
        var d=[];
        var i,b;
        var g=BigInt(count)*BigInt(length);
        for (i=0;i<count;i++) {
            d.push(BigInt("0x"+this.Bin2Hex(bytes.slice(i*sizeD,i*sizeD+sizeD)),16)*g);
        }
        if (bytes.length%count) {
            d[count-1]=BigInt("0x"+this.Bin2Hex(bytes.slice(count*sizeD-sizeD,bytes.length)),16)*g;
        }
        
        var p=parseInt((length-count)/2);
        for (i=0;i<p;i++) {
            d.push(BigInt(i+2));
            for (b=0;b<count;b++) {
                d[d.length-1]=d[b]*d[d.length-1]+g*BigInt(i+1);
            }
        }
        count=d.length;
        var temp=d.slice();
        
        g=this.getHalfInt(g,0)+this.getHalfInt(g,1);
        for (i=0;i<count;i++) {
            for (b=0;b<count;b++) {
                d[b]=d[b]+temp[i]*temp[i]+temp[i]*g;
            }
        }
        var h="",lc=count-1;
        var mc=0;
        while (1) {
            h=this.array_toHex(d);
            if (h.length<length) {
                for (i=0;i<count;i++) {
                    if (i==lc) {
                        d[i]=d[i]+d[i]+g;
                    } else {
                        temp=this.getHalfInt(d[i],0)+this.getHalfInt(d[i],1);
                        d[i]=BigInt(temp.toString()+(this.getHalfInt(d[i+1],0)+this.getHalfInt(d[i+1],1)).toString());
                        d[i+1]=BigInt((this.getHalfInt(d[i+1],0)+this.getHalfInt(d[i+1],1)).toString()+temp.toString());
                    }
                }
            } else if (h.length>length) {
                if (mc>maxIteration) {
                    h=h.substring(0,length);
                    break;
                }
                
                for (i=0;i<count;i++) {
                    if (i==lc) {
                        d[i]=this.getHalfInt(d[i],0)+this.getHalfInt(d[i],1)+g;
                    } else {
                        temp=this.getHalfInt(d[i],1);
                        d[i]=this.getHalfInt(d[i],0)+this.getHalfInt(d[i+1],1);
                        d[i+1]=this.getHalfInt(d[i+1],0)+temp;
                    }
                }
            } else {break;}
            mc++;
        }
        return h;
    }
}*/

class HalHash {
    constructor() {this.version="0.0.2";}
    
    
    Str2Hash(str,length,count=64) {
        return this.Bin2Hash((new TextEncoder("utf-8")).encode(str),length,count);
    }
    
    Bin2Hex(buffer) {
        return [...buffer]
        .map(x => x.toString(16).padStart(2, '0'))
        .join('');
    }
    
    array_toHex(arr) {
        var out="";
        for (var i=0;i<arr.length;i++) {
            out+=arr[i].toString(16);
        }
        return out;
    }
    
    getHalfInt(i,h) {
        var temp=i.toString();
        temp=h==0?temp.slice(0,parseInt(temp.length/2)):temp.slice(parseInt(temp.length/2),temp.length);
        return temp==""?BigInt(0):BigInt(temp);
    }

    Bin2Hash(bytes,length,count=64, maxIteration=100000) {
        //if (bytes.length==1){bytes=new Uint8Array([bytes[0],0]);}
        count=length<count?length:count;
        count=bytes.length<count?bytes.length:count;
        var sizeD=parseInt(bytes.length/count);
        var d=[];
        var i,b;
        var g=BigInt(count)*BigInt(length);
        for (i=0;i<count;i++) {
            d.push(BigInt("0x"+this.Bin2Hex(bytes.slice(i*sizeD,i*sizeD+sizeD)),16)*g);
        }
        if (bytes.length%count) {
            d[count-1]=BigInt("0x"+this.Bin2Hex(bytes.slice(count*sizeD-sizeD,bytes.length)),16)*g;
        }
        
        var p=parseInt((length-count)/2);
        for (i=0;i<p;i++) {
            d.push(BigInt(i+2));
            for (b=0;b<count;b++) {
                d[d.length-1]=d[b]*d[d.length-1]+g*BigInt(i+1);
            }
        }
        count=d.length;
        var temp=d.slice();
        
        g=this.getHalfInt(g,0)+this.getHalfInt(g,1);
        for (i=0;i<count;i++) {
            for (b=0;b<count;b++) {
                d[b]=d[b]+temp[i]*temp[i]+temp[i]*g;
            }
        }
        var h="",lc=count-1;
        var mc=0;
        while (1) {
            h=this.array_toHex(d);
            if (h.length<length) {
                for (i=0;i<count;i++) {
                    if (i==lc) {
                        d[i]=d[i]+d[i]+g;
                    } else {
                        temp=this.getHalfInt(d[i],0)+this.getHalfInt(d[i],1);
                        d[i]=BigInt(temp.toString()+(this.getHalfInt(d[i+1],0)+this.getHalfInt(d[i+1],1)).toString());
                        d[i+1]=BigInt((this.getHalfInt(d[i+1],0)+this.getHalfInt(d[i+1],1)).toString()+temp.toString());
                    }
                }
            } else if (h.length>length) {
                if (mc>maxIteration) {
                    h=h.substring(0,length);
                    break;
                }
                
                for (i=0;i<count;i++) {
                    if (i==lc) {
                        d[i]=this.getHalfInt(d[i],0)+this.getHalfInt(d[i],1)+g;
                    } else {
                        temp=this.getHalfInt(d[i],1);
                        d[i]=this.getHalfInt(d[i],0)+this.getHalfInt(d[i+1],1);
                        d[i+1]=this.getHalfInt(d[i+1],0)+temp;
                    }
                }
            } else {break;}
            mc++;
        }
        return h;
    }
}

File: app/src/main/java/halwarsing/net/halchatandroid/main/HalHash.java
package halwarsing.net.halchatandroid.main;

import android.util.Log;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

//Класс хэширования Халварсинга
public class HalHash {
    private static final byte[] HEX_ARRAY = "0123456789ABCDEF".getBytes(StandardCharsets.US_ASCII);
    private static final String version="0.0.2B";
    private static final String TAG="HH";
    private static final BigInteger TWO=BigInteger.valueOf(2L);
    private static final double LOG2_10 = 0.3010299956639812;
    //0.0.2B - optimized for Android;

    private static final BigInteger[] TEN_POW_CACHE = new BigInteger[500];

    static {
        TEN_POW_CACHE[0]=BigInteger.ONE;
        for(int i=1;i< TEN_POW_CACHE.length;i++) {
            TEN_POW_CACHE[i]=TEN_POW_CACHE[i-1].multiply(BigInteger.TEN);
        }
    }

    public static BigInteger getTenPow(int exp) {
        if(exp>=0&&exp< TEN_POW_CACHE.length) return TEN_POW_CACHE[exp];
        return BigInteger.TEN.pow(exp);
    }

    public static int getDecimalLengthFast(BigInteger number) {
        if(number.equals(BigInteger.ZERO))return 1;
        int digits=(int)(number.bitLength()*LOG2_10)+1;

        if(number.compareTo(getTenPow(digits-1))<0) {
            digits--;
        }
        return digits;
    }


    public String Str2Hash(String str,int length,int count,long maxIteration) {
        return this.Bin2Hash(str.getBytes(StandardCharsets.UTF_8),length,count,maxIteration);
    }

    public String Str2Hash(String str,int length,int count) {
        return Str2Hash(str,length,count,100000);
    }

    public String Str2Hash(String str,int length) {
        return Str2Hash(str,length,64,100000);
    }

    public String Bin2Hex(byte[] bytes) {
        byte[] hexChars = new byte[bytes.length * 2];
        for (int j = 0; j < bytes.length; j++) {
            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = HEX_ARRAY[v >>> 4];
            hexChars[j * 2 + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars, StandardCharsets.UTF_8);
    }

    public String array_toHex(List<BigInteger> arr) {
        StringBuilder out=new StringBuilder();
        for (int i=0;i<arr.size();i++) {
            out.append(arr.get(i).toString(16));
        }
        return out.toString();
    }

    public static int calculateHexLength(List<BigInteger> arr) {
        int totalLength = 0;

        for (BigInteger num : arr) {
            if (num.equals(BigInteger.ZERO)) {
                totalLength += 1;
            } else {
                totalLength+=(num.bitLength()+3)/4;
            }
        }

        return totalLength;
    }

    public static int getDecimalLength(BigInteger number) {
        BigInteger ten = BigInteger.TEN;
        int length = 1;

        while (number.compareTo(ten) >= 0) {
            number = number.divide(ten);
            length++;
        }

        return length;
    }

    public BigInteger[] getHalfInt(BigInteger i) {

        if (i.equals(BigInteger.ZERO)) {
            return new BigInteger[]{BigInteger.ZERO, BigInteger.ZERO};
        }

        int digits=getDecimalLengthFast(i);
        int halfDigits = (digits+1) / 2;

        BigInteger divisor = getTenPow(halfDigits);
        return i.divideAndRemainder(divisor);
    }

    public String Bin2Hash(byte[] bytes,int length,int count,long maxIteration) {
        count= Math.min(length, count);
        count= Math.min(bytes.length, count);
        int sizeD=bytes.length/count;
        List<BigInteger> d=new ArrayList<>();
        BigInteger tempi,tempb;
        int i;
        BigInteger g=BigInteger.valueOf(count).multiply(BigInteger.valueOf(length));

        for (i = 0; i < count; i++) {
            BigInteger segment = new BigInteger(1, Arrays.copyOfRange(bytes, i * sizeD, (i + 1) * sizeD));
            d.add(segment.multiply(g));
        }
        if (bytes.length % count > 0) {
            int start = count * sizeD - sizeD;
            BigInteger remainingSegment = new BigInteger(1, Arrays.copyOfRange(bytes, start, bytes.length));
            d.set(count - 1, remainingSegment.multiply(g));
        }

        int p=(length-count)/2;
        int b;
        for (i=0;i<p;i++) {
            d.add(BigInteger.valueOf(i+2));
            tempi=g.multiply(BigInteger.valueOf(i+1));
            for (b=0;b<count;b++) {
                d.set(d.size()-1,d.get(b).multiply(d.get(d.size()-1)).add(tempi));
            }
        }
        count=d.size();
        List<BigInteger> tempa=new ArrayList<>(d);

        BigInteger[] halfs0,halfs1;
        halfs0=this.getHalfInt(g);
        g=halfs0[0].add(halfs0[1]);

        for (i=0;i<count;i++) {
            tempb=tempa.get(i);
            tempi=tempb.multiply(tempb).add(tempb.multiply(g));
            for (b=0;b<count;b++) {
                d.set(b,d.get(b).add(tempi));
            }
        }

        int lc=count-1;
        long mc=0;
        int hl;

        while (true) {
            hl=HalHash.calculateHexLength(d);
            if (hl < length) {
                for (i = 0; i < count; i++) {
                    if (i == lc) {
                        tempi=d.get(i);
                        d.set(i, tempi.multiply(TWO).add(g));
                    } else {
                        tempi=d.get(i);
                        tempb=d.get(i+1);
                        halfs0=this.getHalfInt(tempi);
                        halfs1=this.getHalfInt(tempb);

                        BigInteger valV=halfs0[0].add(halfs0[1]);
                        BigInteger valZ=halfs1[0].add(halfs1[1]);

                        int lenV=getDecimalLengthFast(valV);
                        int lenZ=getDecimalLengthFast(valZ);

                        BigInteger concatVZ=valV.multiply(getTenPow(lenZ)).add(valZ);
                        BigInteger concatZV=valZ.multiply(getTenPow(lenV)).add(valV);

                        d.set(i,concatVZ);
                        d.set(i+1,concatZV);
                    }
                }
            } else if (hl > length) {
                if (mc > maxIteration) {
                    return this.array_toHex(d).substring(0, length);
                }
                for (i = 0; i < count; i++) {
                    if (i == lc) {
                        tempi=d.get(i);
                        halfs0=this.getHalfInt(tempi);
                        d.set(i, halfs0[0].add(halfs0[1]).add(g));
                    } else {
                        tempi=d.get(i);
                        tempb=d.get(i+1);
                        halfs0=this.getHalfInt(tempi);
                        halfs1=this.getHalfInt(tempb);
                        d.set(i, halfs0[0].add(halfs1[1]));
                        d.set(i + 1, halfs1[0].add(halfs0[1]));
                    }
                }
            } else {
                break;
            }
            mc++;
        }

        return array_toHex(d);
    }

    public String Bin2Hash(byte[] bytes,int length) {
        return Bin2Hash(bytes,length,64,100000);
    }
}
